File: src/Java/643.ArrayFindMaxAverage.java

import java.awt.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.Queue;
import java.util.HashMap;
import java.util.HashSet;

class Solution {
    public double findMaxAverage(int[] nums, int k) {
        int n = nums.length;
        int maxsum = 0;
        for (int i = 0; i < k; i++) {
            maxsum += nums[i];
        }
        int lastsum = maxsum;
        for (int i = k; i < n; i++){
            int nowsum = lastsum - nums[i - k] + nums[i];
            maxsum = Math.max(nowsum, maxsum);
            lastsum = nowsum;
        }
        return maxsum / (double)k;
    }
}

File: src/Python/601-700/643.ArrayFindMaxAverage.py

class Solution:
    def findMaxAverage(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: float
        """
        n = len(nums)
        maxsum = sum(nums[0:k])
        lastsum = maxsum
        for i in range(k, n):
            nowsum = lastsum - nums[i - k] + nums[i]
            lastsum, maxsum = nowsum, max(nowsum, maxsum)
        return maxsum / k

if __name__ == '__main__':
    solution = Solution()
    print(solution.findMaxAverage([1, 12, -5, -6, 50, 3], 4))
    print(solution.findMaxAverage([0, 4, 0, 3, 2], 1))
else:
    pass
